src/andromede/simulation/linearize.py

```diff
@@ -10,138 +10,239 @@
+class ParameterGetter(ABC):
+    @abstractmethod
+    def get_parameter_value(
+        self, component_id: str, parameter_name: str, timestep: int, scenario: int
+    ) -> float:
+        pass
+
+
+@dataclass
+class MutableTerm:
+    coefficient: float
+    component_id: str
+    variable_name: str
+    time_index: int
+    scenario_index: int
+
+    def to_key(self) -> TermKey:
+        return TermKey(
+            self.component_id,
+            self.variable_name,
+            self.time_index,
+            self.scenario_index,
         )
-        result_terms[generate_key(term_with_operator)] = term_with_operator
 
-    # TODO: How can we apply a shift on a parameter ? It seems impossible for now as parameters must already be evaluated...
-    result_expr = LinearExpression(result_terms, input.constant)
-    return result_expr
+    def to_term(self) -> Term:
+        return Term(
+            self.coefficient,
+            self.component_id,
+            self.variable_name,
+            self.time_index,
+            self.scenario_index,
+        )
+
+
+def generate_key(term: MutableTerm) -> TermKey:
+    return TermKey(
+        term.component_id,
+        term.variable_name,
+        term.time_index,
+        term.scenario_index,
+    )
+
+
+@dataclass
+class LinearExpressionData:
+    terms: List[MutableTerm]
+    constant: float
+
+    def build(self) -> LinearExpression:
+        res_terms = {}
+        for t in self.terms:
+            k = generate_key(t)
+            if k in res_terms:
+                current_t = res_terms[k]
+                current_t.coefficient += t.coefficient
+            else:
+                res_terms[k] = t
+        for k, v in res_terms.items():
+            res_terms[k] = v.to_term()
+        return LinearExpression(res_terms, self.constant)
 
 
 @dataclass(frozen=True)
-class LinearExpressionBuilder(ExpressionVisitorOperations[LinearExpression]):
+class LinearExpressionBuilder(ExpressionVisitor[LinearExpressionData]):
     """
     Reduces a generic expression to a linear expression.
 
     Parameters should have been evaluated first.
     """
 
-    structure_provider: IndexingStructureProvider
-    value_provider: Optional[ValueProvider] = None
+    def negation(self, node: NegationNode) -> LinearExpressionData:
+        operand = visit(node.operand, self)
+        operand.constant = -operand.constant
+        for t in operand.terms:
+            t.coefficient = -t.coefficient
+        return operand
+
+    def addition(self, node: AdditionNode) -> LinearExpressionData:
+        operands = [visit(o, self) for o in node.operands]
+        terms = []
+        constant = 0
+        for o in operands:
+            constant += o.constant
+            terms.extend(o.terms)
+        return LinearExpressionData(terms=terms, constant=constant)
+
+    def multiplication(self, node: MultiplicationNode) -> LinearExpressionData:
+        lhs = visit(node.left, self)
```

Comment:
Is there a check somewhere that at least one expression is constant ?

Reply:
Added